### DISCUSSION TITLE
Help with async FFI library

### DISCUSSION THREAD

**Discussion by GuilhermeWerner** [tokio-rs/tokio#3534] [Q&A] [answered]
Hello, I'm a beginner in rust and would like help with a question.

I need to make a library that consumes a json api and I want to provide functions to be called from other applications, via FFI, however there is the problem of tokio::main. For the requests to be asynchronous I need the tokio runtime, but how to use it being that everything will be called in an external application?

It would be unviable to start the runtime on every function call, and it would be great if I didn't need to create my own runtime. I thought about communication between processes, but I think it would not be viable either.

I wonder if anyone has any idea how to do this:

An asynchronous FFI rust library.

**Comment by Darksonn** [comment, marked as answer]
There are many different answers to this question depending on how the non-Rust code needs to talk with the Rust code. The simplest way is to simply create a `tokio::runtime::Runtime` object, and use its `block_on` method to transfer into an async region. By reusing the `Runtime` object, you avoid creation of many new runtimes.

**Reply by GuilhermeWerner** [reply]
Got it, and where would I keep the runtime I created? Because I need it to call the block on event. What would be the best way to reuse the runtime object

**Reply by Darksonn** [reply]
One common way to store Rust objects in C is to create a `Box<RustObj>`, call `Box::into_raw` on it, cast it to a void pointer, then give that void pointer to C. Then when C wants to use it, the Rust methods it calls can cast the pointer back to the right type.

**Reply by GuilhermeWerner** [reply]
I will try, tanks.
